pkg/compactor/block_upload.go

```diff
@@ -294,6 +297,17 @@ func (c *MultitenantCompactor) completeBlockUpload(ctx context.Context, r *http.
 
 	level.Debug(logger).Log("msg", "completing block upload", "files", len(meta.Thanos.Files))
 
+	// Validate blocks by downloading locally and sanitizing
+	rps := newRequestProgressSender(w)
+	if err = rps.SendProgressMessage("starting block validation"); err != nil {
+		return err
+	}
+	if err := c.validateBlock(ctx, w, blockID, userBkt, meta); err != nil {
+		if err = rps.SendProgressMessage(err.Error()); err != nil {
+			return err
+		}
```

Comment:
If validation produced error, we should not only send progress message, but we shoud use `SendFinished(false, ...)` and also return from this method, and not continue with rename of meta file.